### PR TITLE
Re-model the Client

### PR DESCRIPTION
**Pull request by tcharding**
Currently `Client` has the testcontainer in it, as pointed out in review this is slightly odd. We can do better than this. We notice also that this crate is actually useful for more than just testing, currently there are a bunch of helper methods that are test specific spread throughout the crate.

This PR does:

- Split the testcontainer out of `Client` and into its own struct `Container`
- Move all the 'test' helper methods i.e., the alice/bob specific ones into `lib.rs`. 
- Split the integration tests up into client/monerod/wallet that test, respectively, the code in `lib.rs`, `monerod.rs`, and `wallet.rs`

### CODE REVIEW

**Review by bonomat — APPROVED** [review]
I like the result of the refactoring. It's much cleaner now. 

In regards of the usefufness of this library: The purpose of this library is to serve as a tool for testing, that's why it's called monero-harness-rs. If we need it in a different setting, e.g. as a monero-rpc-client, we should structure it differently and extract the functionality in a separate crate. 
Please be thoughtful though: even though it looks useful, it's not proven to be useful nor do we need it in a different setting yet.

**Comment by tcharding** [comment]
> I like the result of the refactoring. It's much cleaner now.
> 
> In regards of the usefufness of this library: The purpose of this library is to serve as a tool for testing, that's why it's called monero-harness-rs. If we need it in a different setting, e.g. as a monero-rpc-client, we should structure it differently and extract the functionality in a separate crate.
> Please be thoughtful though: even though it looks useful, it's not proven to be useful nor do we need it in a different setting yet.

We do actually, we need it for the protocol work @lucas is doing. We need to be able to fetch a transfer based on tx_id, tx_key, and address - the wallet provides the means to do this via this crate. I rekon we should push a full wallet/daemon RPC client implementation upstream into `monero-rs` and just have the 'testing' stuff (alice/bob accounts etc.) in `monero-harness-rs`.

File: src/lib.rs [hunk 1]
```diff
@@ -22,41 +22,81 @@ use rand::Rng;
...
+impl<'c> Container<'c> {
+    /// Creates a new monero container mapping random local ports for the two
+    /// RPC endpoints.
+    pub fn new(cli: &'c Cli) -> Self {
         let mut rng = rand::thread_rng();
         let monerod_rpc_port: u16 = rng.gen_range(1024, u16::MAX);
         let wallet_rpc_port: u16 = rng.gen_range(1024, u16::MAX);
 
-        let container = spin_up_container(cli, monerod_rpc_port, wallet_rpc_port);
+        let image = GenericImage::new("xmrto/monero")
```
Comment (thomaseizinger) [inline src/lib.rs:61]: Instead of using `GenericImage`, you could also just implement the `Image` trait on a struct and pass that to `Docker::run` directly :)

That would obsolete this `Container` struct.

Once you've settled on an interface, you can upstream that to testcontainers then!

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
> > I like the result of the refactoring. It's much cleaner now.
> > In regards of the usefufness of this library: The purpose of this library is to serve as a tool for testing, that's why it's called monero-harness-rs. If we need it in a different setting, e.g. as a monero-rpc-client, we should structure it differently and extract the functionality in a separate crate.
> > Please be thoughtful though: even though it looks useful, it's not proven to be useful nor do we need it in a different setting yet.
> 
> We do actually, we need it for the protocol work @lucas is doing. We need to be able to fetch a transfer based on tx_id, tx_key, and address - the wallet provides the means to do this via this crate. I rekon we should push a full wallet/daemon RPC client implementation upstream into `monero-rs` and just have the 'testing' stuff (alice/bob accounts etc.) in `monero-harness-rs`.

My take on this would be that you have a single workspace in which you build whatever you want to build. In that workspace, you can easily move functionality between crates like splitting out a client, splitting out a test harness, etc.
Most importantly, the thing you are actually trying to build should also be in this workspace. In this case, this would include whatever @luckysori and @rishflab are working on IF you see potential in them using the same functionality.

As you develop and you notice that the interfaces of those crates stabilize and are still providing value, you can move them out into a different repository or upstream them to some other lib.

Upstreaming stuff is cool but doing it prematurely can cause quite some friction in various ways:

- It is often surprising, how you would want interfaces to change to better accommodate for your usecase.
- The upstream maintainers will have their own ideas of how things should be integrated. It will require time and energy to balance these things.
- Sometimes, you end up changing things on such a large scale that what you wanted to upstream is simply no longer needed. If you already upstreamed, you find yourself in a situation where you convinced a maintainer of the usefulness of a piece of functionality, basically talked them into maintaining stuff for you and you end up not actually using it. I've been there, doesn't feel great :sweat_smile:

**Comment by tcharding** [comment]
> > > I like the result of the refactoring. It's much cleaner now.
> > > In regards of the usefufness of this library: The purpose of this library is to serve as a tool for testing, that's why it's called monero-harness-rs. If we need it in a different setting, e.g. as a monero-rpc-client, we should structure it differently and extract the functionality in a separate crate.
> > > Please be thoughtful though: even though it looks useful, it's not proven to be useful nor do we need it in a different setting yet.
> > 
> > 
> > We do actually, we need it for the protocol work @lucas is doing. We need to be able to fetch a transfer based on tx_id, tx_key, and address - the wallet provides the means to do this via this crate. I rekon we should push a full wallet/daemon RPC client implementation upstream into `monero-rs` and just have the 'testing' stuff (alice/bob accounts etc.) in `monero-harness-rs`.
> 
> My take on this would be that you have a single workspace in which you build whatever you want to build. In that workspace, you can easily move functionality between crates like splitting out a client, splitting out a test harness, etc.
> Most importantly, the thing you are actually trying to build should also be in this workspace. In this case, this would include whatever @luckysori and @rishflab are working on IF you see potential in them using the same functionality.
> 
> As you develop and you notice that the interfaces of those crates stabilize and are still providing value, you can move them out into a different repository or upstream them to some other lib.
> 
> Upstreaming stuff is cool but doing it prematurely can cause quite some friction in various ways:
> 
> * It is often surprising, how you would want interfaces to change to better accommodate for your usecase.
> * The upstream maintainers will have their own ideas of how things should be integrated. It will require time and energy to balance these things.
> * Sometimes, you end up changing things on such a large scale that what you wanted to upstream is simply no longer needed. If you already upstreamed, you find yourself in a situation where you convinced a maintainer of the usefulness of a piece of functionality, basically talked them into maintaining stuff for you and you end up not actually using it. I've been there, doesn't feel great sweat_smile

Thanks for the input @thomaseizinger, so paraphrasing to see I got you. 

1. Move `monero-harnes-rs`  into a module within `thor`
2. Give it time to stabalise
3. Split it out into separate crates and/or upstream it after its stabalized

I like that approach.

Cheers

File: src/lib.rs (hunk 1 above)
Comment (tcharding) [reply src/lib.rs:61]: Nice, will do this in a separate PR.

**Comment by thomaseizinger** [comment]
Yes!

One small comment but I think you understood correctly and just worded it ambiguously 😁

Cargo workspaces are super powerful: You can "split out a crate" from some other code without moving it to a separate repository / releasing it.

That is a really powerful in-between state where you can already utilize crates to design boundaries (and have `cargo` enforce them) without having to fully commit to them in terms of releases, versioning etc.

As such, I would rephrase it as:

1. Make the thor repository a cargo workspace
2. Move the crate(s) of this repo into that workspace
3. Set/Refine the crate boundaries as you develop

From there you can upstream things but also start publishing those crates at any point! No need to move them to a separate repository :)
